File: typescript-workshop/part-2-light-mocks/tests/service/DiscountCodeServiceTest.ts
import { describe, it, expect, vi } from 'vitest';
import { DiscountCode } from '../../src/domain/DiscountCode';
import type { DiscountCodeRepository } from '../../src/repository/DiscountCodeRepository';
import { DiscountCodeService } from '../../src/service/DiscountCodeService';

describe('DiscountCodeService', () => {
  it('p2_checkDiscountCode_returnsTrue_whenCodeHasAlreadyBeenUsed', () => {
    const repository = { checkDiscountCode: vi.fn(), markAsUsed: vi.fn() } as unknown as DiscountCodeRepository;
    const discountCodeService = new DiscountCodeService(repository);

    (repository as any).checkDiscountCode.mockReturnValue(true);

    expect(discountCodeService.checkDiscountCode('CUST-001', DiscountCode.SUMMER20)).toBe(true);
  });

  it('p2_checkDiscountCode_returnsFalse_whenCodeHasNotBeenUsed', () => {
    const repository = { checkDiscountCode: vi.fn(), markAsUsed: vi.fn() } as unknown as DiscountCodeRepository;
    const discountCodeService = new DiscountCodeService(repository);

    (repository as any).checkDiscountCode.mockReturnValue(false);

    expect(discountCodeService.checkDiscountCode('CUST-001', DiscountCode.SUMMER20)).toBe(false);
  });

  it('p2_markAsUsed_delegatesToRepository', () => {
    const repository = { checkDiscountCode: vi.fn(), markAsUsed: vi.fn() } as unknown as DiscountCodeRepository;
    const discountCodeService = new DiscountCodeService(repository);

    discountCodeService.markAsUsed('CUST-001', DiscountCode.SUMMER20);

    expect((repository as any).markAsUsed).toHaveBeenCalledWith('CUST-001', DiscountCode.SUMMER20);
  });
});

File: csharp-workshop/Part2.LightMocks.Tests/Service/DiscountCodeServiceTest.cs
using NSubstitute;
using Part2.LightMocks.Domain;
using Part2.LightMocks.Repository;
using Part2.LightMocks.Service;
using Xunit;

namespace Part2.LightMocks.Tests.Service;

public class DiscountCodeServiceTest
{
    [Fact]
    public void P2_CheckDiscountCode_ReturnsTrue_WhenCodeHasAlreadyBeenUsed()
    {
        var repository = Substitute.For<IDiscountCodeRepository>();
        var discountCodeService = new DiscountCodeService(repository);

        repository.CheckDiscountCode("CUST-001", DiscountCode.SUMMER20).Returns(true);

        Assert.True(discountCodeService.CheckDiscountCode("CUST-001", DiscountCode.SUMMER20));
    }

    [Fact]
    public void P2_CheckDiscountCode_ReturnsFalse_WhenCodeHasNotBeenUsed()
    {
        var repository = Substitute.For<IDiscountCodeRepository>();
        var discountCodeService = new DiscountCodeService(repository);

        repository.CheckDiscountCode("CUST-001", DiscountCode.SUMMER20).Returns(false);

        Assert.False(discountCodeService.CheckDiscountCode("CUST-001", DiscountCode.SUMMER20));
    }

    [Fact]
    public void P2_MarkAsUsed_DelegatesToRepository()
    {
        var repository = Substitute.For<IDiscountCodeRepository>();
        var discountCodeService = new DiscountCodeService(repository);

        discountCodeService.MarkAsUsed("CUST-001", DiscountCode.SUMMER20);

        repository.Received(1).MarkAsUsed("CUST-001", DiscountCode.SUMMER20);
    }
}
